fix: abstract type support for `inject()`

JanUnld opened the pull request:
- added `AbstractType` and `ProviderToken`
- added `InjectOptions` to `inject()`
- added optional support for `inject()`

fixes #68

JanUnld committed:
fix: introduced abstract type support for injections

- added `AbstractType` and `ProviderToken`
- added `InjectOptions` to `inject()`
- added optional support for `inject()`

mgechev commented:
@JanUnld sorry for the late response. Would you add tests for the change? I'll merge and release.

JanUnld commented:
@mgechev no worries! Thanks for taking a look at my PRs.

Tests are added and updated. I've also introduced support for optional injection when using `inject()` and added proper typing for multi provider support in this PR.

mgechev commented:
Released in v2.6.1.
